### PR TITLE
Count down notification regarding issue-11172

### LINKED ISSUE #11172: Visiting top level private message list should not clear all PM notifications
[//]: # (Before logging this issue, please post to the Report a Bug guild from the Habitica website's Help menu. Most bugs can be handled quickly there. If a GitHub issue is needed, you will be advised of that by a moderator or staff member -- a player with a dark blue or purple name. It is recommended that you don't create a new issue unless advised to.)

[//]: # (Bugs in the mobile apps can also be reported there.)

[//]: # (If you have a feature request, use "Help > Request a Feature", not GitHub or the Report a Bug guild.)

[//]: # (For more guidelines see https://github.com/HabitRPG/habitica/issues/2760)

[//]: # (Fill out relevant information - UUID is found from the Habitia website at User Icon > Settings > API)
### Description
[//]: # (Describe bug in detail here. Include screenshots if helpful.)

AnneLine in the Report a Bug Guild outlined this usability struggle:

>Not sure if this is a bug or intended behavior. Sometimes I get 3 or 4 PM messages at a time. When I come on Habitica, I see a flag saying that. Nice! But once I have gone into the box to check, there is no longer a flag for the messages, even if I only replied to one. Several times, I've not noticed I had not read nor responded to messages that way. Is it behaving as intended, or could it be made to count down the way the guild notifications do?
>
> May be just something that's an issue for me, but not for others, but here's where the problem happens. I look at one PM from the App (which triggers the turn off of the number), both that one AND any others that might be in there... but I'm not on a computer, and/or I'm only focused on doing that one thing that made me go there.... But maybe it's something that has to be done from the computer rather than the app. Or I need to fetch some info before replying. Now I might remember THAT task... but there's no longer a 'flag' for me that there might be another message or two hiding in there... and it might be a day or two longer till I get an additional message... and HOPEFULLY see other PMs and realize I din't read them and didn't respond.
>
>Could it possible to set up be something like a toggle to allow to continue to show additional messages or something? This is probably veering very close to a request rather than a bug - but it did delay a response to Alys about something for a few days.....

This workflow difficulty is worth addressing. If we kept the notification bubble live until each individual unread conversation were visited, that'd help ensure that users don't miss any notifications while processing their private messages.

### PR DESCRIPTION
**Pull request by aevix**
Sorry for taking so long to come up with the solution for this issue, I am still pretty new to coding and it took me a hot minute to figure out how the frontend and backend were talking to each other. I got it working and I am sure there are things I missed but this is what I have so far.

I want to give some of the credit to @ResamVi. I missed that he already submitted a PR few months ago regarding this issue and I wish I saw his PR earlier because that could have saved me a few days of research. His PR had a lot of the things solved already that I took and tweaked to work with the new develop version. 

Before:
When a sender sends a message to the receiver, it creates two messages, one for the sender and one for the receiver. It also increments the `store.state.user.data.inbox.newMessage` for the receiver per message. When user goes to the messages under user dropdown menu, it calls the api v3 mark_pms_read which resets the `inbox.newMessages` counter to 0 and refreshes the frontend.

[//]: # (Note: See http://habitica.fandom.com/wiki/Using_Your_Local_Install_to_Modify_Habitica%27s_Website_and_API for more info)

Fixes #11172 

### Changes
[//]: # Starting from the backend, I added a read property to the inbox schema and set it to false on default. When user sends the message, in the methods.js file I added a read property set as true for the sender's message in the sender's inbox so that it doesn't count the message as unread later in the chain. I also added an if statement that basically resets the counter to 0 if for some reason the counter goes to a negative number. I noticed during testing when counter became negative it was triggering all sorts of weird issues for notification and that might have some relations to issue #11199. 
In the private-messages.vue file, under `loadMessages` method, I added the function that @ResamVi had in his code that looks for any messages that were loaded that had the `read`property as false and saves it in `unreadMessagesCount` object. It emits a function call for `markPrivMessagesRead` which updates the `newMessages` counter on the frontend side. Also api is called using `axios.get` and passes `unreadMessagesCount` and `toUserUuid `to backend in api-v4/inbox.js file. This api takes the request and passes it to `markPmsRead` from inbox/index.js which subtracts the `unreadMessagesCount` from `user.inbox.newMessages` and saves the state and changes the message read property to true. 

----
UUID: 228f739f-b42c-4801-932c-10b4d739b460

### CODE REVIEW

**Comment by aevix** [comment]
I changed mark-pms-read test so it passes but I am not sure why its failing on the send-private-message test. I don't think the changes I made would have affected any of the api for sending messages.

File: test/api/v3/integration/user/POST-user_mark_pms_read.test.js [hunk 1]
```diff
@@ -1,11 +1,11 @@
 import {
   generateUser,
-} from '../../../../helpers/api-integration/v3';
+} from '../../../../helpers/api-integration/v4';
 
-describe('POST /user/mark-pms-read', () => {
+describe('GET /user/mark-pms-read', () => {
```
Comment (paglias) [inline test/api/v3/integration/user/POST-user_mark_pms_read.test.js:5]: Looks like this is a breaking change which we can't have. Any existing route should keep working as it's now, basically you can add something new but not break an existing functionality

File: test/api/v3/integration/user/POST-user_mark_pms_read.test.js [hunk 2]
```diff
@@ -1,11 +1,11 @@
...
   let user;
 
-  beforeEach(async () => {
+  before(async () => {
```
Comment (paglias) [inline test/api/v3/integration/user/POST-user_mark_pms_read.test.js:8]: why was this changed?

**Comment by paglias** [comment]
Hi @aevix , I'm very sorry for the delay. I think this approach might be a bit too complex as it requires changes to the existing clients (web & mobile) and also changes the existing data structure for PMs in the database. I think something simpler would be to add a new route (or a parameter to the existing one) that when called reduces the number of unread pms by 1

**Comment by aevix** [comment]
@paglias I will try and come up with a simpler solution and let you know.

**Comment by aevix** [comment]
@paglias I've been thinking about a way to just use the api already set up to simply subtract the notification by 1 every time the user opens a new message but the problem is that since the notification isn't set up to track individual messages, it just stores the static number of new messages received in the receiver's inbox when sender sends their message, so if user opens a specific new message two or three times it will continue to subtract from the newMessage counter even though there are still messages that were not read. Hopefully I am making sense.

I don't see any work around this issue unless we add something to the inbox schema to keep track of every message whether it is read or not and show notification for only messages that are marked not read. There might be a easier method that I am not seeing but this seems to be the only solution I can come up with. I would like to hear your input to maybe clue me into something I am not seeing.

Thanks!

File: website/server/models/message.js [hunk 3]
```diff
@@ -45,6 +45,7 @@ const inboxSchema = new mongoose.Schema({
   // we store two copies of each inbox messages:
   // one for the sender and one for the receiver
   ownerId: { $type: String, ref: 'User' },
+  read: { $type: Boolean, default: false },
```
Comment (paglias) [inline website/server/models/message.js:48]: My main concern with this one is that we'll have to run an huge migration to mark as read all existing messages and we actually won't know if they've been read or not

**Comment by paglias** [comment]
@aevix 

> @paglias I've been thinking about a way to just use the api already set up to simply subtract the notification by 1 every time the user opens a new message but the problem is that since the notification isn't set up to track individual messages, it just stores the static number of new messages received in the receiver's inbox when sender sends their message, so if user opens a specific new message two or three times it will continue to subtract from the newMessage counter even though there are still messages that were not read. Hopefully I am making sense.

That's a good point, hmm.. I left a few comments but I don't think there's an easy solution and changing the schema, apart from being a little complex right now, has the problem that we don't know which PMs have been read at the moment so we won't know if they have to be marked as read or unread in the database

**Comment by aevix** [comment]
> 

> @aevix
> 
> > @paglias I've been thinking about a way to just use the api already set up to simply subtract the notification by 1 every time the user opens a new message but the problem is that since the notification isn't set up to track individual messages, it just stores the static number of new messages received in the receiver's inbox when sender sends their message, so if user opens a specific new message two or three times it will continue to subtract from the newMessage counter even though there are still messages that were not read. Hopefully I am making sense.
> 
> That's a good point, hmm.. I left a few comments but I don't think there's an easy solution and changing the schema, apart from being a little complex right now, has the problem that we don't know which PMs have been read at the moment so we won't know if they have to be marked as read or unread in the database

since we don't have any info on whether message is read or not now, couldn't we just make default as false and then create a button that marks all messages as read? Although with this solution I am not sure if that would cause any issues on the backend server side with potentially thousands of requests at once changing each message as read but that could be one solution. 

Another solution i thought was do two updates where initially mark every message as read on migration and then do a second update to the schema to make default to false so any new messages that are created after the second update will be correctly marked. I am not sure if either of these would be a good solution but it would be interesting to hear your thoughts.

File: test/api/v3/integration/user/POST-user_mark_pms_read.test.js (hunk 1 above)
Comment (aevix) [reply test/api/v3/integration/user/POST-user_mark_pms_read.test.js:5]: Okay, I will have to go back and create a separate test for the new api and revert all the changes. Ill update it once I am done and get it working.

File: test/api/v3/integration/user/POST-user_mark_pms_read.test.js (hunk 2 above)
Comment (aevix) [reply test/api/v3/integration/user/POST-user_mark_pms_read.test.js:8]: I don't think that should have been changed, it might have been a typo on my part. Ill go ahead and fix it.

**Comment by paglias** [comment]
> Another solution i thought was do two updates where initially mark every message as read on migration and then do a second update to the schema to make default to false so any new messages that are created after the second update will be correctly marked. I am not sure if either of these would be a good solution but it would be interesting to hear your thoughts.

But you can't know which of the existing messages have been read or if they're still to be read

**Comment by aevix** [comment]
@paglias I don't know if there is a way to figure out if a message is read since current inbox schema doesn't track that information. We could maybe add a function for users to mark a message as unread if they want to go back manually.

**Comment by SabreCat** [comment]
Closing per author's comment here: https://github.com/HabitRPG/habitica/issues/11172#issuecomment-784700783

This may need some discussion between UX and technical devs, ultimately, before we open it back up for contributors. Thank you for the effort diving into this, though, @aevix!
